File: app/schemas/thesis.py
"""Pydantic contract for the 7-pillar investment thesis (BUILD_PLAN.md §4).

Adapted from the original spec's JSON contract with the three additions
BUILD_PLAN.md calls for: classification.currency, structured fields inside
what_can_kill_it[], and outcome. See harness/memory/decisions.md for the
adaptation notes (human-readable enum strings, metric-key re-keying) made
against the user-supplied Balu Forge sample.
"""
from __future__ import annotations

from datetime import date
from typing import Literal, Optional
from urllib.parse import urlparse

from pydantic import BaseModel, Field, field_validator, model_validator, ValidationInfo

OPERATING_MODELS = {"factory", "subscription", "money_lending", "retail_stores", "services"}
THESIS_STATUSES = {"on_track", "watch_closely", "broken"}


def _normalize_enum(value: str) -> str:
    return value.strip().lower().replace(" ", "_").replace("-", "_")


def _strip_markdown_link(value: str) -> str:
    import re

    match = re.match(r"^\[.*?\]\((.*?)\)$", value.strip())
    return match.group(1) if match else value.strip()


class RevenueSplitItem(BaseModel):
    segment: str
    share_pct: float = Field(ge=0, le=100)


class TheBusiness(BaseModel):
    what_it_does: str
    revenue_split: list[RevenueSplitItem]

    @model_validator(mode="after")
    def _split_sums_to_100(self) -> "TheBusiness":
        total = sum(item.share_pct for item in self.revenue_split)
        if abs(total - 100) > 0.5:
            raise ValueError(f"revenue_split.share_pct sums to {total}, expected 100 ± 0.5")
        return self


class TheBigChange(BaseModel):
    summary: str
    expected_completion: str


class ProofPoints(BaseModel):
    hard_evidence: list[str] = Field(default_factory=list)
    model_specific_metrics: dict[str, float] = Field(default_factory=dict)


class KillTrigger(BaseModel):
    label: str
    metric_key: Optional[str] = None
    operator: Optional[Literal["<", "<=", ">", ">=", "==", "!="]] = None
    threshold: Optional[float] = None
    unit: Optional[str] = None
    action: str
    severity: Literal["warn", "kill"] = "kill"
    grace_periods: int = Field(default=1, ge=1)
    manual_check: bool = False

    @model_validator(mode="after")
    def _quantifiable_unless_manual(self) -> "KillTrigger":
        if not self.manual_check:
            if self.metric_key is None or self.operator is None or self.threshold is None:
                raise ValueError(
                    "kill trigger must set manual_check=true, or provide "
                    "metric_key, operator, and threshold"
                )
        return self


class HealthCheckHistoryItem(BaseModel):
    quarter: str
    verdict: str
    note: str


class HealthCheckPillar(BaseModel):
    latest_quarter_review: str
    historical_checks: list[HealthCheckHistoryItem] = Field(default_factory=list)


class ReferenceItem(BaseModel):
    title: str
    url: str

    @field_validator("url", mode="before")
    @classmethod
    def _strip_and_validate_url(cls, value: str) -> str:
        cleaned = _strip_markdown_link(value)
        parsed = urlparse(cleaned)
        if parsed.scheme not in ("http", "https") or not parsed.netloc:
            raise ValueError(f"references[].url is not a valid absolute URL: {value!r}")
        return cleaned


class ThesisData(BaseModel):
    the_business: TheBusiness
    the_growth_engine: list[str] = Field(default_factory=list)
    the_big_change: TheBigChange
    proof_points: ProofPoints
    what_can_kill_it: list[KillTrigger]
    why_we_believe_it: list[str]
    health_check: HealthCheckPillar
    references: list[ReferenceItem] = Field(default_factory=list)

    @model_validator(mode="after")
    def _at_least_one_kill_severity(self) -> "ThesisData":
        if not any(t.severity == "kill" for t in self.what_can_kill_it):
            raise ValueError("what_can_kill_it must contain at least one entry with severity='kill'")
        return self

    @model_validator(mode="after")
    def _why_we_believe_it_shape(self) -> "ThesisData":
        entries = self.why_we_believe_it
        if len(entries) < 3:
            raise ValueError(f"why_we_believe_it needs >= 3 entries, got {len(entries)}")
        premises = [e for e in entries if e.strip().lower().startswith("premise")]
        conclusions = [e for e in entries if e.strip().lower().startswith("conclusion")]
        if len(premises) < 1:
            raise ValueError("why_we_believe_it must contain at least one entry starting with 'Premise'")
        if len(conclusions) != 1:
            raise ValueError(
                f"why_we_believe_it must contain exactly one 'Conclusion' entry, found {len(conclusions)}"
            )
        return self

    @model_validator(mode="after")
    def _metric_keys_in_registry(self, info: ValidationInfo) -> "ThesisData":
        context = info.context or {}
        registry = context.get("metric_registry")
        operating_model = context.get("operating_model")
        if registry is None:
            return self  # registry not supplied — caller (service layer / test) opted out

        used_keys: list[str] = list(self.proof_points.model_specific_metrics.keys())
        used_keys += [t.metric_key for t in self.what_can_kill_it if t.metric_key]

        for key in used_keys:
            if key not in registry:
                raise ValueError(f"unknown metric_key '{key}': not present in metric_definitions")
            key_model = registry[key]  # None means universal
            if key_model is not None and key_model != operating_model:
                raise ValueError(
                    f"metric_key '{key}' belongs to operating_model '{key_model}', "
                    f"not this company's '{operating_model}'"
                )
        return self


class Classification(BaseModel):
    broad_industry: str
    specific_niche: str
    operating_model: str
    currency: str = Field(default="INR", pattern=r"^[A-Z]{3}$")

    @field_validator("operating_model", mode="before")
    @classmethod
    def _normalize_operating_model(cls, value: str) -> str:
        normalized = _normalize_enum(value)
        if normalized not in OPERATING_MODELS:
            raise ValueError(f"unknown operating_model '{value}'; expected one of {sorted(OPERATING_MODELS)}")
        return normalized


class ThesisCreate(BaseModel):
    company_id: str = Field(pattern=r"^[A-Z0-9_]{2,50}$")
    name: str
    classification: Classification
    status: str = "on_track"
    last_reviewed: date
    thesis_data: ThesisData

    @field_validator("status", mode="before")
    @classmethod
    def _normalize_status(cls, value: str) -> str:
        normalized = _normalize_enum(value)
        if normalized not in THESIS_STATUSES:
            raise ValueError(f"unknown status '{value}'; expected one of {sorted(THESIS_STATUSES)}")
        return normalized

File: web/src/lib/server/schemas/thesis.ts
// Ports app/schemas/thesis.py's Pydantic contract to Zod.
import { z } from 'zod';
import { PILLAR_KEYS } from '../pillars';

const THESIS_STATUSES = ['on_track', 'watch_closely', 'broken'] as const;

const normalizeEnum = (v: string) => v.trim().toLowerCase().replace(/ /g, '_').replace(/-/g, '_');
const stripMarkdownLink = (v: string) => {
	const m = v.trim().match(/^\[.*?\]\((.*?)\)$/);
	return m ? m[1] : v.trim();
};

export const revenueSplitItem = z.object({
	segment: z.string(),
	share_pct: z.number().min(0).max(100)
});

// A thesis gets built up incrementally - nothing in it is required to save.
// revenue_split summing to ~100 is only checked when the analyst has actually
// entered a full split (empty is fine, a partial one they're mid-typing is
// left alone rather than blocked).
export const theBusiness = z
	.object({
		what_it_does: z.string().default(''),
		revenue_split: z.array(revenueSplitItem).default([])
	})
	.refine((v) => v.revenue_split.length === 0 || Math.abs(v.revenue_split.reduce((s, i) => s + i.share_pct, 0) - 100) <= 0.5, {
		message: 'revenue_split.share_pct must sum to 100 +/- 0.5 once entries are added'
	});

export const theBigChange = z.object({
	summary: z.string().default(''),
	expected_completion: z.string().default('')
});

export const proofPoints = z.object({
	hard_evidence: z.array(z.string()).default([]),
	model_specific_metrics: z.record(z.string(), z.number()).default({})
});

// A half-filled trigger row (label typed, threshold not yet decided) is a
// normal mid-draft state, not an error - only reject a trigger that claims to
// be metric-driven (manual_check=false) but sets just some of the three
// metric fields, since that combination can't evaluate to anything.
export const killTrigger = z
	.object({
		label: z.string().default(''),
		metric_key: z.string().nullable().optional(),
		operator: z.enum(['<', '<=', '>', '>=', '==', '!=']).nullable().optional(),
		threshold: z.number().nullable().optional(),
		unit: z.string().nullable().optional(),
		action: z.string().default(''),
		severity: z.enum(['warn', 'kill']).default('kill'),
		grace_periods: z.number().int().min(1).default(1),
		manual_check: z.boolean().default(false)
	})
	.refine(
		(v) => {
			if (v.manual_check) return true;
			const set = [v.metric_key, v.operator, v.threshold].filter((x) => x != null);
			return set.length === 0 || set.length === 3;
		},
		{ message: 'kill trigger metric_key, operator, and threshold must all be set together (or leave all three blank)' }
	);

export const healthCheckHistoryItem = z.object({
	quarter: z.string(),
	verdict: z.string(),
	note: z.string()
});

export const healthCheckPillar = z.object({
	latest_quarter_review: z.string().default(''),
	historical_checks: z.array(healthCheckHistoryItem).default([])
});

export const referenceItem = z.object({
	title: z.string().default(''),
	url: z
		.string()
		.default('')
		.transform(stripMarkdownLink)
		.refine(
			(v) => {
				if (!v) return true;
				try {
					const u = new URL(v);
					return u.protocol === 'http:' || u.protocol === 'https:';
				} catch {
					return false;
				}
			},
			{ message: 'references[].url must be a valid absolute URL when provided' }
		)
});

export const thesisData = z
	.object({
		the_business: theBusiness.default({ what_it_does: '', revenue_split: [] }),
		the_growth_engine: z.array(z.string()).default([]),
		the_big_change: theBigChange.default({ summary: '', expected_completion: '' }),
		proof_points: proofPoints.default({ hard_evidence: [], model_specific_metrics: {} }),
		what_can_kill_it: z.array(killTrigger).default([]),
		why_we_believe_it: z.array(z.string()).default([]),
		health_check: healthCheckPillar.default({ latest_quarter_review: '', historical_checks: [] }),
		trackables: z.array(z.string().refine((value) => !!value.trim(), 'Trackable cannot be blank')).default([]),
		buy_sell_decision: z.string().default(''),
		references: z.array(referenceItem).default([]),
		pillar_notes: z.record(z.string(), z.array(z.string())).default({})
	})
	.superRefine((v, ctx) => {
		// The only thing still checked here is internal consistency of data
		// that *was* provided (an unknown pillar_notes key is a bug, not an
		// incomplete thesis) - nothing about the thesis being incomplete is an
		// error. A thesis builds up over time; every pillar can be empty and
		// filled in later.
		const unknown = Object.keys(v.pillar_notes).filter((k) => !(PILLAR_KEYS as readonly string[]).includes(k));
		if (unknown.length) {
			ctx.addIssue({
				code: 'custom',
				message: `pillar_notes has unknown key(s) ${JSON.stringify(unknown)}; must be one of ${JSON.stringify(PILLAR_KEYS)}`
			});
		}
	});

export type ThesisData = z.infer<typeof thesisData>;

export const classification = z.object({
	broad_industry: z.string(),
	specific_niche: z.string(),
	// operating_model used to be checked against a fixed list here; it's now
	// an open, user-extensible lookup table (operating_models) - existence is
	// validated in versioning.ts's resolveOperatingModel, same place
	// broad_industry/specific_niche get validated against their tables.
	operating_model: z.string().min(1).transform(normalizeEnum),
	currency: z
		.string()
		.regex(/^[A-Z]{3}$/)
		.default('INR')
});

const tickerField = z
	.string()
	.trim()
	.toUpperCase()
	.regex(/^[A-Z0-9&.-]{1,20}$/, 'must be a short alphanumeric ticker/scrip code')
	.optional()
	.or(z.literal(''))
	.transform((v) => (v ? v : undefined));

export const thesisCreate = z
	.object({
		// company_id is optional on input - createCompany() derives it from
		// whichever ticker is present when omitted, since the form no longer
		// asks the user to type an internal ID directly (see ingest/+page.svelte).
		company_id: z
			.string()
			.regex(/^[A-Z0-9_]{2,50}$/)
			.optional(),
		nse_ticker: tickerField,
		bse_ticker: tickerField,
		name: z.string(),
		classification,
		status: z.string().transform((v, ctx) => {
			const n = normalizeEnum(v);
			if (!(THESIS_STATUSES as readonly string[]).includes(n)) {
				ctx.addIssue({ code: 'custom', message: `unknown status '${v}'` });
				return z.NEVER;
			}
			return n as (typeof THESIS_STATUSES)[number];
		}).default('on_track'),
		// Defaults to today rather than requiring the analyst to pick a date
		// just to save a still-empty thesis.
		last_reviewed: z.string().default(() => new Date().toISOString().slice(0, 10)),
		thesis_data: thesisData.default(() => thesisData.parse({}))
	})
	.superRefine((v, ctx) => {
		if (!v.nse_ticker && !v.bse_ticker && !v.company_id) {
			ctx.addIssue({ code: 'custom', message: 'at least one of nse_ticker or bse_ticker is required' });
		}
	});

export type ThesisCreate = z.infer<typeof thesisCreate>;
